### PR TITLE
[Default Envs] Add default conda environment to Sklearn

### PR DESCRIPTION
**Pull request by dbczumar**
This PR is a the first in a multi-part series that adds default Conda environments to each MLflow model flavor and attempts to standardize logic between existing flavors.

Followup PRs:

- A PR adding default environments for the PyTorch, Tensorflow, and Spark flavors with unit tests. The changes for each flavor will be very similar to the `sklearn` changes.

- A PR adding default environments for the h2o and Keras flavors with unit tests. The changes for each flavor will be very similar to the `sklearn` changes.

- A PR adding integration tests for sagemaker deployment and pyfunc serving with default environments for each flavor

### CODE REVIEW

File: mlflow/utils/model_utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import os
+
+from mlflow.exceptions import MlflowException
+from mlflow.models import Model
+from mlflow.protos.databricks_pb2 import RESOURCE_DOES_NOT_EXIST
+
+def _get_flavor_configuration(model_path, flavor_name):
```
Comment (mparkhe) [inline mlflow/utils/model_utils.py]: Can you add unit test coverage for this function? Error and passing conditions

File: mlflow/utils/environment.py [hunk 2]
```diff
@@ -9,15 +9,18 @@
 """
 
 
-def _mlflow_conda_env(path, additional_conda_deps=None, additional_pip_deps=None,
-                      additional_conda_channels=None):
+def _mlflow_conda_env(path=None, additional_conda_deps=None, additional_pip_deps=None,
+        additional_conda_channels=None):
```
Comment (mparkhe) [inline mlflow/utils/environment.py]: **nit** : indent

File: mlflow/utils/environment.py [hunk 3]
```diff
@@ -9,15 +9,18 @@
...
     """
     Create conda environment file. Contains default dependency on current python version.
-    :param path: local filesystem path where the conda env file is to be created.
+    :param path: Local filesystem path where the conda env file is to be created. If unspecified,
+                 the conda env will be returned as a string.
     :param additional_conda_deps: List of additional conda dependencies passed as strings.
     :param additional_pip_deps: List of additional pip dependencies passed as strings.
     :param additional_channels: List of additional conda channels to search when resolving packages.
-    :return: path where the conda environment file has been created.
+    :return: Either:
```
Comment (mparkhe) [inline mlflow/utils/environment.py]: Can we avoid this non-uniformity of returned values. For instance, if the client has specified `path` -- does that call stack need to be returned that same string back?

File: mlflow/utils/environment.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/utils/environment.py]: Fixed!

File: mlflow/utils/environment.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/utils/environment.py]: Done! We always return the environment text, regardless of whether or not the environment was written to the filesystem.

File: mlflow/utils/model_utils.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/utils/model_utils.py]: Added a `tests/utils/test_model_utils.py` suite with tests for the error and passing conditions associated with this method.

**Review by dbczumar — COMMENTED** [review]
@mparkhe Thanks for the review. Addressed your comments!

File: mlflow/sklearn.py [hunk 4]
```diff
@@ -22,6 +22,18 @@
 from mlflow.models import Model
 from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE, INTERNAL_ERROR
 import mlflow.tracking
+from mlflow.utils.environment import _mlflow_conda_env
+from mlflow.utils.model_utils import _get_flavor_configuration
+
+FLAVOR_NAME = "sklearn"
+
+DEFAULT_CONDA_ENV = _mlflow_conda_env(
```
Comment (tomasatdatabricks) [inline mlflow/sklearn.py]: I would maybe turn this into a function taking a path and returning path instead. Then you don't need to change environemt._mlflow_conda_env and I think global function is preferable to global string

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good, thanks Corey. I left a few comments.

My main question is whether it makes sense to change the _mlflow_conda_env to return string instead of path. I think in most case you want the path so it's a more convenient interface. But I am ok if you think  returning text is better, it's just a suggestion.

File: mlflow/utils/environment.py [hunk 5]
```diff
@@ -28,6 +32,16 @@ def _mlflow_conda_env(path, additional_conda_deps=None, additional_pip_deps=None
...
+    env_text = yaml.safe_dump(env, stream=None, default_flow_style=False)
+
+    if path is not None:
+        with open(path, "w") as f:
+            f.write(env_text)
+
+    return env_text
+
+
+def _get_base_env():
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/environment.py]: Nope! Thanks for catching that

**Comment by dbczumar** [comment]
@tomasatdatabricks I removed the unused `_get_base_env` function and responded to your comments. Thanks for the review!

File: mlflow/utils/environment.py [hunk 6]
```diff
@@ -9,15 +9,19 @@
...
+    Creates a Conda environment with the specified package channels and dependencies.
+    Optionally, saves the Conda environment to an output path, if specified.
+
+    :param path: Local filesystem path where the conda env file is to be written. If unspecified,
+                 the conda env will not be written to the filesystem; it will still be returned
+                 in string format.
     :param additional_conda_deps: List of additional conda dependencies passed as strings.
     :param additional_pip_deps: List of additional pip dependencies passed as strings.
     :param additional_channels: List of additional conda channels to search when resolving packages.
-    :return: path where the conda environment file has been created.
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/environment.py]: After discussing offline with @tomasatdatabricks, we've decided to have two return cases for this function:

1. Return `None` if an output path is specified
2. Return the yaml representation of the environment as a dictionary if an output path is NOT specified.

File: mlflow/sklearn.py (hunk 4 above)
Comment (tomasatdatabricks) [reply mlflow/sklearn.py]: Discussed offline, looks good now!

**Review by tomasatdatabricks — APPROVED** [review]
lgtm

**Pull request merged**